pyquil/api/quantum_computer.py

```diff
@@ -0,0 +1,183 @@
+from math import pi
+from typing import List
+
+import numpy as np
+
+from pyquil.api._qam import QAM
+from pyquil.device import AbstractDevice
+from pyquil.gates import MEASURE, RX
+from pyquil.quil import Program, get_classical_addresses_from_program
+from pyquil.quilbase import Measurement, Pragma
+
+
+def _get_flipped_protoquil_program(program: Program):
+    """For symmetrization, generate a program where X gates are added before measurement.
+
+    Forest 1.3 is really picky about where the measure instructions happen. It has to be
+    at the end!
+    """
+    program = Program(program.instructions)  # Copy
+    to_measure = []
+    while True:
+        inst = program.instructions[-1]
+        if isinstance(inst, Measurement):
+            program.pop()
+            to_measure.append((inst.qubit, inst.classical_reg))
+        else:
+            break
+
+    program += Pragma('PRESERVE_BLOCK')
+    for qu, addr in to_measure[::-1]:
+        program += RX(pi, qu)
+    program += Pragma('END_PRESERVE_BLOCK')
+
+    for qu, addr in to_measure[::-1]:
+        program += Measurement(qubit=qu, classical_reg=addr)
+
+    return program
+
+
+class QuantumComputer:
+    def __init__(self, *, name: str, qam: QAM, device: AbstractDevice, symmetrize=False):
+        """
+        A quantum computer for running quantum programs.
+
+        A quantum computer has various characteristics like supported gates, qubits, qubit
+        topologies, gate fidelities, and more. A quantum computer also has the ability to
+        run quantum programs.
+
+        A quantum computer can be a real Rigetti QPU that uses superconducting transmon
+        qubits to run quantum programs, or it can be an emulator like the Rigetti QVM with
+        noise models and mimicked topologies.
+
+        :param name: A string identifying this particular quantum computer.
+        :param qam: A quantum abstract machine which handles executing quantum programs. This
+            dispatches to a QVM or QPU.
+        :param device: A collection of connected qubits and associated specs and topology.
+        :param symmetrize: Whether to apply readout error symmetrization. See
+            :py:func:`run_symmetrized` for a complete description.
+        """
+        self.name = name
+        self.qam = qam
+        self.device = device
+
+        self.symmetrize = symmetrize
+
+    def qubit_topology(self):
+        return self.device.qubit_topology()
+
+    def get_isa(self, oneq_type='Xhalves', twoq_type='CZ'):
+        return self.device.get_isa(oneq_type=oneq_type, twoq_type=twoq_type)
+
+    def run(self, program, classical_addresses, trials, symmetrize=None) -> np.ndarray:
```

Comment:
Are people going to be clear on the difference between QuantumComputer.run and QPU.run?

Reply:
Probably not ... 

In my mind, QPU.run and QVM.run typify the `QAM` interface. This sets the minimal collection of methods that an abstract quantum machine needs to satisfy to work with pyquil. Currently it's just `run`. You can imagine slotting in additional QAM's like the referenceqvm work or other simulators. You can also imagine evolving the interface to have additional (but still "low level") methods.

`QuantumComputer` includes the qam, a description of the device (real or otherwise) and a bunch of convenience methods for executing quantum programs that build on top of `QAM.run`. If you add a new method (e.g. `sampling_expectation(prog, pauli)`) it instantly works with anything that satisfies the `QAM` interface. So there's no actual difference between `QPU.run` and `QuantumComputer.run`, but there *is* a difference between `QPU` and `QuantumComputer`

1) does this make any sense?
2) do you have suggestions for how to convey this to users?